simde/x86/gfni.h

```diff
@@ -0,0 +1,765 @@
+  #define simde_mm256_maskz_gf2p8affineinv_epi64_epi8(k, x, A, b) _mm256_maskz_gf2p8affineinv_epi64_epi8(k, x, A, b)
+#else
+  SIMDE__FUNCTION_ATTRIBUTES
+  simde__m256i
+  simde_mm256_maskz_gf2p8affineinv_epi64_epi8 (simde__mmask32 k, simde__m256i x, simde__m256i A, int b) {
+    simde__m256i r = simde_mm256_gf2p8affineinv_epi64_epi8(x, A, b);
+    simde__m256i_private r_ = simde__m256i_to_private(r);
+
+    SIMDE__VECTORIZE
+    for (size_t i = 0 ; i < (sizeof(r_.u8) / sizeof(r_.u8[0])) ; i++) {
+      r_.u8[i] = ((k >> i) & 1) ? r_.u8[i] : 0;
+    }
+
+    return simde__m256i_from_private(r_);
+  }
+  #if defined(SIMDE_GFNI_ENABLE_NATIVE_ALIASES)
+    #define _mm256_maskz_gf2p8affineinv_epi64_epi8(k, x, A, b) simde_mm256_maskz_gf2p8affineinv_epi64_epi8(k, x, A, b)
+  #endif
+#endif
+
+#if defined(SIMDE_GFNI_NATIVE) && defined(SIMDE_AVX512F_NATIVE)
+  #define simde_mm512_maskz_gf2p8affineinv_epi64_epi8(k, x, A, b) _mm512_maskz_gf2p8affineinv_epi64_epi8(k, x, A, b)
+#else
+  SIMDE__FUNCTION_ATTRIBUTES
+  simde__m512i
+  simde_mm512_maskz_gf2p8affineinv_epi64_epi8 (simde__mmask64 k, simde__m512i x, simde__m512i A, int b) {
+    simde__m512i r = simde_mm512_gf2p8affineinv_epi64_epi8(x, A, b);
+    simde__m512i_private r_ = simde__m512i_to_private(r);
+
+    SIMDE__VECTORIZE
+    for (size_t i = 0 ; i < (sizeof(r_.u8) / sizeof(r_.u8[0])) ; i++) {
+      r_.u8[i] = ((k >> i) & 1) ? r_.u8[i] : 0;
+    }
+
+    return simde__m512i_from_private(r_);
+  }
+  #if defined(SIMDE_GFNI_ENABLE_NATIVE_ALIASES)
+    #define _mm512_maskz_gf2p8affineinv_epi64_epi8(k, x, A, b) simde_mm512_maskz_gf2p8affineinv_epi64_epi8(k, x, A, b)
+  #endif
+#endif
+
+SIMDE__FUNCTION_ATTRIBUTES
+simde__m128i simde_mm_gf2p8mul_epi8 (simde__m128i a, simde__m128i b) {
+  #if defined(SIMDE_GFNI_NATIVE) && defined(SIMDE_AVX512VL_NATIVE)
+    return _mm_gf2p8mul_epi8(a, b);
+  #elif defined(SIMDE_ARCH_X86_SSE2)
+    simde__m128i r, t, zero, ones, fgp;
+    zero = simde_mm_setzero_si128();
+    r = zero;
+    ones = simde_mm_set1_epi8(0x01);
+
+    /* The field generator polynomial is 0x11B but we drop the 0x100 bit */
+    fgp = simde_mm_set1_epi8(0x1B);
+
+    for (int i = 0; i < 8; ++i) {
+      t = simde_mm_and_si128(b, ones);
+      t = simde_mm_cmpeq_epi8(t, ones);
+      t = simde_mm_and_si128(a, t);
+      r = simde_mm_xor_si128(r, t);
+      t = simde_mm_cmpgt_epi8(zero, a);
+      t = simde_mm_and_si128(fgp, t);
+      a = simde_mm_add_epi8(a, a);
+      a = simde_mm_xor_si128(a, t);
+      b = simde_mm_srli_epi64(b, 1);
+    }
+
+    return r;
+  #else
+    simde__m128i_private
+      r_,
+      a_ = simde__m128i_to_private(a),
+      b_ = simde__m128i_to_private(b);
+
+    SIMDE__VECTORIZE
+    for (size_t i = 0 ; i < (sizeof(r_.u8) / sizeof(r_.u8[0])) ; i++) {
+      r_.u8[i] = 0;
+      while ((a_.u8[i] != 0) && (b_.u8[i] != 0)) {
+	if (b_.u8[i] & 1)
+	  r_.u8[i] ^= a_.u8[i];
+
+	if (a_.u8[i] & 0x80)
+	  a_.u8[i] = (a_.u8[i] << 1) ^ 0x11B;
```

Comment:
Microsoft Visual C on arm reports:
> warning C4244: '=': conversion from 'int' to 'uint8_t', possible loss of data

Reply:
I'll correct these.
Are these other compilers free (as in beer)?
I didn't think that SIMDe was endian-safe so I haven't paid any attention to endianism.
Unfortunately I don't have a big-endian machine (apart maybe from digging out an old Linksys NSLU2 which IIRC ran Debian armbe).
However I doubt it would be up to the task :-(

Reply:
`icc` is free as in beer. You can see the CI setup at https://github.com/nemequ/simde/blob/master/.travis.yml#L120

Microsoft Visual Studio appears to be "freemium": https://en.wikipedia.org/wiki/Microsoft_Visual_Studio#Editions

If you have Debian than you can install `gcc-9-s390x-linux-gnu` and `g++-9-s390x-linux-gnu` or similar for quick compilation along with `qemu-user-static` for running the tests on s390x

Reply:
Thanks for the information.
Silly me not to think about using qemu for big-endian tests.
I already use it to run ARM code on my x86 hardware.

Reply:
> However I doubt it would be up to the task :-(

Ugh, yeah, don't do that.  Michael's right about qemu on Debian being the right way to go for this (https://wiki.debian.org/Multiarch/HOWTO is valuable information).

Usually the endian issues are really problems where we use the wrong types internally (i.e., treating an `int16_t` as 2 × `int8_t`.  That said, I don't really grok you code yet (I haven't looked at it *that* closely), so I can't promise that's the issue.

What OS/distro do you use?  https://software.intel.com/en-us/articles/oneapi-repo-instructions has instructions for various options… If you're on OS X I don't think there is a good option right now without applying to Intel for a license for open source development (not a big deal, I've had one for years, but https://devcloud.intel.com/oneapi/ would work.

All the compilers we test on CI can be installed for free.  In addition to VS Express, there is also [PGI Community Edition](https://www.pgroup.com/products/community.htm), and XL C/C++ is available for free on POWER machines (on x86 it costs money) and I can provide you access to a very fast POWER VM for SIMDe development courtesy of [IntegriCloud](https://integricloud.com/).  I may even be able to get a BE distro running on there if that would help, I haven't checked yet.